Add a config parameter CONNECTION_TIMEOUT

xavierfacq opened the pull request:
Add a config parameter CONNECTION_TIMEOUT to set the maximum connection timeout when performing HTTP requests.

dgiagio commented on src/main/java/com/suse/saltstack/netapi/config/ClientConfig.java:
```diff
@@ -12,6 +12,9 @@
     public static final Key<String> TOKEN = new Key<>();
 
+    // Timeout settings setted at 1000ms (10s)
+    public static final Key<Integer> CONNECTION_TIMEOUT = new Key<>(1000);
```
Good job on the patch. But isn't 1000ms equal to 1s instead of 10s?

xavierfacq replied:
Thanks, there was a mistake in my units, it's commited.

dgiagio commented:
@xavierfacq Ok, good. Can you also add a unit test for the timeout? The WireMock framework has some options to delay the response, so you can ensure it's working the way it's supposed to.

lucidd commented:
@xavierfacq Thanks for your Contribution! 
Would you mind implementing the timeout also for the `JDKConnection` backend? [Here](http://docs.oracle.com/javase/7/docs/api/java/net/URLConnection.html#setConnectTimeout(int)) is the relevant documentation.

xavierfacq commented:
@dgiagio I'll add the testunit ASAP
@lucidd Is the JDKConnection implementation ok for you ?

lucidd commented:
@xavierfacq Im not sure yet i will have to look into it more. I think the best case would be to have the same config keys used by both back ends and have them behave the same to avoid confusion. I'm not sure if this is possible tho (also considering possibly having additional back ends in the future). What do you think would that be possible?

xavierfacq commented:
@lucidd You are right, I'll simplify and add only a new key named REQUEST_CONNECTION_TIMEOUT,   used by both back ends.

lucidd commented:
I looked into both connection back ends a bit and it seems that there are 2 types of timeouts that both back ends can provide. 

            | JDK          | HttpClient
-----------|------------- | -------------
initial connection | [setConnectTimeout](http://docs.oracle.com/javase/7/docs/api/java/net/URLConnection.html#setConnectTimeout(int))  | [setConnectTimeout](http://hc.apache.org/httpcomponents-client-ga/httpclient/apidocs/org/apache/http/client/config/RequestConfig.html#getConnectTimeout())
waiting for data | [setReadTimeout](http://docs.oracle.com/javase/7/docs/api/java/net/URLConnection.html#setReadTimeout(int))  | [setSocketTimeout](http://hc.apache.org/httpcomponents-client-ga/httpclient/apidocs/org/apache/http/client/config/RequestConfig.html#getSocketTimeout())

Note: i linked the getters for `HttpClient` since only they have documentation.

So i would suggest having 2 config keys `CONNECT_TIMEOUT` and `SOCKET_TIMEOUT` that are used by both back ends.

xavierfacq commented:
Ok, I'll omit the setConnectionRequestTimeout() and only set the two main timeouts : 
- opening a communications link : using setConnectTimeout()
- waiting for data : using setReadTimeout() / setSocketTimeout()

Ok for your config key names ;-)

lucidd commented on src/test/java/com/suse/saltstack/netapi/client/SaltStackClientTest.java:
```diff
@@ -146,6 +148,23 @@ public void testRunRequestAsync() throws Exception {
 
+    @Test(expected = SaltStackException.class)
+    public void testRunRequestWithTimeout() throws Exception {
+    	// create a local SaltStackClient with a fast timeout configuration to do not lock tests more thant 2s
```
here is a tab that tried to sneak in ;)

lucidd commented on src/main/java/com/suse/saltstack/netapi/config/ClientConfig.java:
```diff
@@ -12,6 +12,25 @@
+    // Request timeout settings
+    /**
+     * Sets a specified timeout value, in milliseconds, to be used when dealing a communications link in HttpClientConnection / JDKConnection.
+     * A timeout of zero is interpreted as an infinite timeout. Default value is 10000ms (10s).
```
I would write something more similar to the HttpClient documentation:
`Timeout in milliseconds until a connection is established. A timeout of zero is interpreted as an infinite timeout. Default value is 10000ms (10s)`

lucidd commented on src/main/java/com/suse/saltstack/netapi/config/ClientConfig.java:
```diff
@@ -12,6 +12,25 @@
+
+    /**
+     * Sets a specified timeout value, in milliseconds, to be used when waiting for data in HttpClientConnection / JDKConnection.
+     * A timeout of zero is interpreted as an infinite timeout. Default value is 10000ms (10s).
```
And here too:
`Timeout in milliseconds for waiting for data. A timeout of zero is interpreted as an infinite timeout. Default value is 10000ms (10s)`

lucidd commented:
@xavierfacq We should test both timeouts separately and with both back ends. You can pass the back end to use to the `SaltStackClient`.
That said i haven't found a way to test `CONNECT_TIMEOUT` yet since `withFixedDelay`
does not work in that case. Maybe @dgiagio knows more.

xavierfacq commented:
I have changed the Javadoc and added the following text : `A negative value is interpreted as undefined (system default).` I have modified the code of JDKConnection to deal with negative values, to be compliant with HttpClient.

I'm looking for a solution to test the CONNECT_TIMEOUT in SaltStackClientTest, and I'll create tests for JDKConnection.

renner commented on src/main/java/com/suse/saltstack/netapi/config/ClientConfig.java:
```diff
@@ -12,6 +12,25 @@
     public static final Key<URI> URL = new Key<>(URI.create("http://localhost:8000"));
     public static final Key<String> TOKEN = new Key<>();
 
+    // Request timeout settings
```
I don't think we need this single line comment since there is the javadoc one below.

renner commented on src/main/java/com/suse/saltstack/netapi/config/ClientConfig.java:
```diff
@@ -12,6 +12,25 @@
+    public static final Key<Integer> CONNECT_TIMEOUT = new Key<>(10000);
+
+    /**
+     * Timeout in milliseconds for waiting for data. A timeout of zero is interpreted as an infinite timeout. A negative value is interpreted as undefined (system default).
```
Please follow the maximum line length convention (92 characters) also in javadoc comments, here and above.

isbm commented on src/main/java/com/suse/saltstack/netapi/client/impl/JDKConnection.java:
```diff
@@ -70,6 +73,13 @@ private T request(String method, String data)
 
+            // Timeouts may be specified on configuration
+            int connect_timeout = config.get(CONNECT_TIMEOUT);
+            if (connect_timeout >= 0) connection.setConnectTimeout(connect_timeout);
```
No, please don't.
http://www.oracle.com/technetwork/java/javase/documentation/codeconventions-142311.html

isbm replied:
Also, I am not sure if we want to trim negative here. If `config` is expecting to return 0, then it setting timeout to zero is same as disabling it, so you can just set it directly from getting it. But if configuration has negative value, this should be as configuration error, while this information is lost at this point.

But I don't have a strong opinion here. @renner, @moio ?

isbm commented on src/main/java/com/suse/saltstack/netapi/client/impl/JDKConnection.java:
```diff
@@ -70,6 +73,13 @@ private T request(String method, String data)
+            if (connect_timeout >= 0) connection.setConnectTimeout(connect_timeout);
+
+            int socket_timeout = config.get(SOCKET_TIMEOUT);
+            if (socket_timeout >= 0) connection.setReadTimeout(socket_timeout);
```
Add braces here as well.

isbm commented on src/main/java/com/suse/saltstack/netapi/config/ClientConfig.java:
```diff
@@ -12,6 +12,28 @@
+     * @see HttpClientConnection.request(String)
+     * @see JDKConnection.request(String, String)
+     */
+    public static final Key<Integer> SOCKET_TIMEOUT = new Key<>(10000);
```
Please move `10000` to a `DEFAULT_TIMEOUT` variable.

renner replied:
I would say this is not necessary and can stay as it is and @moio agrees with that as well.

lucidd replied:
I don't think we need extra constants for the default values of keys since they are already pretty self explanatory constants accessible through `SOCKET_TIMEOUT.defaultValue`.
Besides that they may have the same default value now but that could change.

isbm commented on src/main/java/com/suse/saltstack/netapi/config/ClientConfig.java:
```diff
@@ -12,6 +12,28 @@
+     * @see HttpClientConnection.request(String)
+     * @see JDKConnection.request(String, String)
+     */
+    public static final Key<Integer> CONNECT_TIMEOUT = new Key<>(10000);
```
Please see comment for line 35.

xavierfacq replied:
Is it ok for you if I add a `private static final int DEFAULT_TIMEOUT = 10000;` used to initialize `CONNECT_TIMEOUT` and `SOCKET_TIMEOUT` ?

isbm replied:
Why constant `DEFAULT_TIMEOUT` has to be private?

xavierfacq replied:
There are already two public config keys, a third one controlling the two others will be strange, no ?

isbm replied:
It is already strange we have so many default timeouts, is it?

isbm commented on src/main/java/com/suse/saltstack/netapi/config/ClientConfig.java:
```diff
@@ -12,6 +12,28 @@
+
     // Proxy settings
     public static final Key<String> PROXY_HOSTNAME = new Key<>();
     public static final Key<Integer> PROXY_PORT = new Key<>(3128);
```
If this port is not a random idea, I would suggest to move it to constants instead.

renner replied:
This port (3128) was chosen as the default PROXY_PORT since it is the default port used by some HTTP proxy servers, e.g. squid. Making this a constant? Why not, but no need to discuss this here since it is very unrelated to this patch.

isbm replied:
@renner agreed.

isbm commented on src/test/java/com/suse/saltstack/netapi/client/SaltStackClientTest.java:
```diff
@@ -147,6 +155,46 @@ public void testRunRequestAsync() throws Exception {
+        // to do not lock tests more thant 2s
+        URI uri = URI.create("http://localhost:" + Integer.toString(MOCK_HTTP_PORT));
+        SaltStackClient clientWithFastTimeout = new SaltStackClient(uri);
+        clientWithFastTimeout.getConfig().put(SOCKET_TIMEOUT, 1000);
```
Bug. You use default timeout 10 seconds, but here only one second.

renner replied:
Why is this a bug @isbm? He suggests to use 1 second here (and below) to not block the tests for 10 seconds. Makes sense to me actually.

isbm replied:
Because you do not test a default timeout? :-) In this case you can put 50 ms as well, as a positive non-zero will set a timeout as per se.

isbm commented on src/test/java/com/suse/saltstack/netapi/client/SaltStackClientTest.java:
```diff
@@ -147,6 +155,46 @@ public void testRunRequestAsync() throws Exception {
+        // to do not lock tests more thant 2s
+        URI uri = URI.create("http://localhost:" + Integer.toString(MOCK_HTTP_PORT));
+        SaltStackClient clientWithFastTimeout = new SaltStackClient(uri, new JDKConnectionFactory());
+        clientWithFastTimeout.getConfig().put(SOCKET_TIMEOUT, 1000);
```
Same here, I guess. I would strongly suggest to use `DEFAULT_TIMEOUT` constant instead.

xavierfacq replied:
This two tests check that the "Read timed out" is well implemented, that's why I force a short timeout of 1s, and delay the server response of 2s, to catch and valid the "Read timed out" error. This is just an hack to quickly pass tests. If we want to test with the default config, we have to delay the response more than 10000ms, twice... 

Is it OK ?

renner replied:
I agree and from my perspective this test is OK as it is using the 1 second timeout. It could in theory even be smaller as @isbm suggested, but IMO 1 second is small enough. The default value could maybe be verified just like this (if this is even necessary)?

    assertEquals(10000, new SaltStackClient().getConfig().get(SOCKET_TIMEOUT));

renner commented:
I merged it, thanks for your work so far! Please have a look at the fixes that I added on top of your commits, especially those two:

- Give proper names to variables (7bc990ed5bcfe44e44c471303cef791ddc777640)
- Fix typos in comments (9be2854917fbeac0d46087ea6cca646a7e904981)

xavierfacq commented:
I agree for your fixes, thanks for the merge !
